File: miniapp/src/api/nutrition.ts
/**
 * API-слой модуля питания: типы, запросы, React Query хуки.
 * Все запросы идут через apiClient с авторизацией X-Init-Data.
 */
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'

// ── Типы ──────────────────────────────────────────────────────────────────────

/** Продукт из справочника (КБЖУ на 100г) */
export interface FoodItem {
  id: number
  name: string
  calories: number | null
  protein_g: number | null
  fat_g: number | null
  carbs_g: number | null
  serving_size_g?: number | null
  is_favorite?: boolean
}

/** Позиция в приёме пищи (КБЖУ на порцию) */
export interface MealItem {
  id: number
  food_item_id: number
  name: string
  amount_g: number
  calories: number
  protein_g: number
  fat_g: number
  carbs_g: number
}

/** Приём пищи с позициями и итогами */
export interface Meal {
  id: number
  meal_type: 'breakfast' | 'lunch' | 'dinner' | 'snack'
  eaten_at: string | null
  notes: string
  items: MealItem[]
  total_calories: number
  total_protein: number
  total_fat: number
  total_carbs: number
  created_at: string | null
}

/** Цели по питанию */
export interface NutritionGoal {
  calories: number | null
  protein_g: number | null
  fat_g: number | null
  carbs_g: number | null
  serving_size_g?: number | null
  water_ml: number
  goal_type: string | null        // lose / maintain / gain
  activity_level: string | null   // sedentary / light / moderate / active / very_active
}

/** Параметры тела пользователя */
export interface Profile {
  weight_kg: number | null
  height_cm: number | null
  age: number | null
  gender: string | null   // male / female
}

/** DTO обновления профиля */
export interface ProfileUpdateDto {
  weight_kg?: number | null
  height_cm?: number | null
  age?: number | null
  gender?: string | null
}

/** Запрос авто-расчёта целей */
export interface GoalsCalculateRequest {
  goal_type: string
  activity_level: string
  weight_kg: number
  height_cm: number
  age: number
  gender: string
  water_ml?: number
}

/** Результат авто-расчёта */
export interface GoalsCalculateResponse {
  calories: number
  protein_g: number
  fat_g: number
  carbs_g: number
  water_ml: number
}

/** Суммарные КБЖУ за день */
export interface Totals {
  calories: number
  protein_g: number
  fat_g: number
  carbs_g: number
}

/** Полная сводка за день */
export interface DaySummary {
  date: string
  goals: NutritionGoal | null
  totals: Totals
  water_ml: number
  meals: Meal[]
}

/** Статистика за период */
export interface Stats {
  period: string
  days: number
  avg_calories: number
  avg_protein: number
  avg_fat: number
  avg_carbs: number
  avg_water: number
  daily: DaySummary[]
}

/** Шаблон приёма пищи */
export interface MealTemplate {
  id: number
  name: string
  meal_type: string
  items: MealItem[]
  total_calories: number
  created_at: string | null
}

/** DTO создания шаблона из приёма */
export interface TemplateFromMealDto {
  meal_id: number
  name: string
}

/** Ответ на запрос предложений */
export interface SuggestionsResponse {
  frequent_foods: (FoodItem & { use_count: number })[]
  recent_meals: Meal[]
}

/** DTO позиции при создании приёма пищи */
export interface MealItemCreate {
  name: string
  amount_g: number
  calories: number
  protein_g: number
  fat_g: number
  carbs_g: number
}

/** DTO создания приёма пищи */
export interface MealCreateDto {
  meal_type: string
  eaten_at?: string
  items: MealItemCreate[]
  notes?: string
}

/** DTO обновления целей */
export interface GoalUpdateDto {
  calories?: number | null
  protein_g?: number | null
  fat_g?: number | null
  carbs_g?: number | null
  water_ml?: number | null
  goal_type?: string | null
  activity_level?: string | null
}

// ── API-функции ───────────────────────────────────────────────────────────────

/** Сводка за сегодня */
const fetchToday = async (): Promise<DaySummary> =>
  (await apiClient.get('/nutrition/today')).data

/** Приёмы пищи за период */
const fetchMeals = async (dateFrom: string, dateTo: string): Promise<Meal[]> =>
  (await apiClient.get('/nutrition/meals', { params: { date_from: dateFrom, date_to: dateTo } })).data

/** Один приём пищи */
export const fetchMeal = async (id: number): Promise<Meal> =>
  (await apiClient.get(`/nutrition/meal/${id}`)).data

/** Создание приёма пищи */
const createMeal = async (dto: MealCreateDto): Promise<Meal> =>
  (await apiClient.post('/nutrition/meal', dto)).data

/** Обновление позиций приёма пищи */
const updateMeal = async ({ id, items }: { id: number; items: MealItemCreate[] }): Promise<Meal> =>
  (await apiClient.put(`/nutrition/meal/${id}`, { items })).data

/** Удаление приёма пищи */
const deleteMealApi = async (id: number): Promise<void> =>
  await apiClient.delete(`/nutrition/meal/${id}`)

/** Логирование воды */
const logWaterApi = async (amount_ml: number): Promise<{ id: number; amount_ml: number; logged_at: string }> =>
  (await apiClient.post('/nutrition/water', { amount_ml })).data

/** Вода за дату */
const fetchWater = async (date: string): Promise<{ date: string; total_ml: number }> =>
  (await apiClient.get('/nutrition/water', { params: { date } })).data

/** Текущие цели */
const fetchGoals = async (): Promise<NutritionGoal> =>
  (await apiClient.get('/nutrition/goals')).data

/** Обновление целей */
const updateGoalsApi = async (dto: GoalUpdateDto): Promise<NutritionGoal> =>
  (await apiClient.put('/nutrition/goals', dto)).data

/** Список шаблонов */
const fetchTemplates = async (): Promise<MealTemplate[]> =>
  (await apiClient.get('/nutrition/templates')).data

/** Создать шаблон из приёма пищи */
const createTemplateFromMealApi = async (dto: TemplateFromMealDto): Promise<MealTemplate> =>
  (await apiClient.post('/nutrition/templates/from-meal', dto)).data

/** Удалить шаблон */
const deleteTemplateApi = async (id: number): Promise<void> =>
  await apiClient.delete(`/nutrition/templates/${id}`)

/** Применить шаблон (создать meal) */
const applyTemplateApi = async (id: number): Promise<Meal> =>
  (await apiClient.post(`/nutrition/templates/${id}/apply`)).data

/** Предложения (частые продукты + последние приёмы) */
const fetchSuggestions = async (mealType?: string): Promise<SuggestionsResponse> =>
  (await apiClient.get('/nutrition/suggestions', { params: mealType ? { meal_type: mealType } : {} })).data

/** Список избранных продуктов */
const fetchFavorites = async (): Promise<FoodItem[]> =>
  (await apiClient.get('/nutrition/favorites')).data

/** Toggle избранного */
const toggleFavoriteApi = async (foodItemId: number): Promise<{ food_item_id: number; is_favorite: boolean }> =>
  (await apiClient.post(`/nutrition/favorites/${foodItemId}`)).data

/** Получить профиль (параметры тела) */
const fetchProfile = async (): Promise<Profile> =>
  (await apiClient.get('/nutrition/profile')).data

/** Обновить профиль */
const updateProfileApi = async (dto: ProfileUpdateDto): Promise<Profile> =>
  (await apiClient.put('/nutrition/profile', dto)).data

/** Авто-расчёт целей по параметрам тела */
export const calculateGoalsApi = async (dto: GoalsCalculateRequest): Promise<GoalsCalculateResponse> =>
  (await apiClient.post('/nutrition/goals/calculate', dto)).data

/** Поиск продуктов */
const searchFoodApi = async (q: string): Promise<FoodItem[]> =>
  (await apiClient.get('/nutrition/food/search', { params: { q } })).data

/** Статистика за период */
const fetchStats = async (period: 'week' | 'month'): Promise<Stats> =>
  (await apiClient.get('/nutrition/stats', { params: { period } })).data

// ── React Query хуки ──────────────────────────────────────────────────────────

/** Сводка за сегодня */
export function useNutritionToday() {
  return useQuery({ queryKey: ['nutrition', 'today'], queryFn: fetchToday })
}

/** Сводка за конкретную дату */
export function useNutritionDay(date: string) {
  return useQuery({
    queryKey: ['nutrition', 'day', date],
    queryFn: async () => {
      // Используем meals + water + goals для произвольной даты
      const [meals, water, goals] = await Promise.all([
        fetchMeals(date, date),
        fetchWater(date),
        fetchGoals(),
      ])
      // Рассчитываем итоги
      const totals: Totals = {
        calories: meals.reduce((s, m) => s + m.total_calories, 0),
        protein_g: meals.reduce((s, m) => s + m.total_protein, 0),
        fat_g: meals.reduce((s, m) => s + m.total_fat, 0),
        carbs_g: meals.reduce((s, m) => s + m.total_carbs, 0),
      }
      return { date, goals, totals, water_ml: water.total_ml, meals } as DaySummary
    },
    enabled: !!date,
  })
}

/** Приёмы пищи за период */
export function useMeals(dateFrom: string, dateTo: string) {
  return useQuery({
    queryKey: ['nutrition', 'meals', dateFrom, dateTo],
    queryFn: () => fetchMeals(dateFrom, dateTo),
    enabled: !!dateFrom && !!dateTo,
  })
}

/** Создание приёма пищи */
export function useCreateMeal() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: createMeal,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition'] })
    },
  })
}

/** Обновление приёма пищи */
export function useUpdateMeal() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: updateMeal,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition'] })
    },
  })
}

/** Удаление приёма пищи */
export function useDeleteMeal() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: deleteMealApi,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition'] })
    },
  })
}

/** Логирование воды */
export function useLogWater() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: logWaterApi,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition'] })
    },
  })
}

/** Цели */
export function useGoals() {
  return useQuery({ queryKey: ['nutrition', 'goals'], queryFn: fetchGoals })
}

/** Обновление целей */
export function useUpdateGoals() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: updateGoalsApi,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition', 'goals'] })
      qc.invalidateQueries({ queryKey: ['nutrition', 'today'] })
    },
  })
}

/** Список шаблонов */
export function useTemplates() {
  return useQuery({ queryKey: ['nutrition', 'templates'], queryFn: fetchTemplates })
}

/** Создать шаблон из приёма пищи */
export function useCreateTemplateFromMeal() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: createTemplateFromMealApi,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition', 'templates'] })
    },
  })
}

/** Удалить шаблон */
export function useDeleteTemplate() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: deleteTemplateApi,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition', 'templates'] })
    },
  })
}

/** Применить шаблон */
export function useApplyTemplate() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: applyTemplateApi,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition'] })
    },
  })
}

/** Предложения (частые + последние) */
export function useSuggestions(mealType?: string) {
  return useQuery({
    queryKey: ['nutrition', 'suggestions', mealType],
    queryFn: () => fetchSuggestions(mealType),
    staleTime: 60_000,
  })
}

/** Избранные продукты */
export function useFavorites() {
  return useQuery({ queryKey: ['nutrition', 'favorites'], queryFn: fetchFavorites })
}

/** Toggle избранного */
export function useToggleFavorite() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: toggleFavoriteApi,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition', 'favorites'] })
      qc.invalidateQueries({ queryKey: ['nutrition', 'food', 'search'] })
    },
  })
}

/** Профиль (параметры тела) */
export function useProfile() {
  return useQuery({ queryKey: ['nutrition', 'profile'], queryFn: fetchProfile })
}

/** Обновление профиля */
export function useUpdateProfile() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: updateProfileApi,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['nutrition', 'profile'] })
    },
  })
}

/** Авто-расчёт целей */
export function useCalculateGoals() {
  return useMutation({ mutationFn: calculateGoalsApi })
}

/** Поиск продуктов с debounce */
export function useFoodSearch(query: string) {
  return useQuery({
    queryKey: ['nutrition', 'food', 'search', query],
    queryFn: () => searchFoodApi(query),
    enabled: query.length >= 2,
    staleTime: 30_000,
  })
}

/** Статистика за период */
export function useNutritionStats(period: 'week' | 'month') {
  return useQuery({
    queryKey: ['nutrition', 'stats', period],
    queryFn: () => fetchStats(period),
  })
}

File: api/routers/nutrition.py
"""
REST API роутер для модуля питания (/api/nutrition).
Все операции изолированы по user_id текущего пользователя.
"""
from __future__ import annotations

from datetime import date, datetime, timezone
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field

from api.deps import get_current_user
from db.models import User
from db import nutrition_storage as ns

router = APIRouter(prefix="/nutrition", tags=["nutrition"])


# ── Pydantic-схемы ────────────────────────────────────────────────────────────

class FoodItemOut(BaseModel):
    """Продукт из справочника (КБЖУ на 100г)."""
    id: int
    name: str
    calories: Optional[float] = None
    protein_g: Optional[float] = None
    fat_g: Optional[float] = None
    carbs_g: Optional[float] = None
    serving_size_g: Optional[float] = None


class MealItemOut(BaseModel):
    """Позиция в приёме пищи (КБЖУ рассчитаны на порцию)."""
    id: int
    food_item_id: int
    name: str
    amount_g: float
    calories: float
    protein_g: float
    fat_g: float
    carbs_g: float


class MealOut(BaseModel):
    """Приём пищи с позициями и итогами."""
    id: int
    meal_type: str
    eaten_at: Optional[str] = None
    notes: str = ""
    items: List[MealItemOut] = []
    total_calories: float = 0
    total_protein: float = 0
    total_fat: float = 0
    total_carbs: float = 0
    created_at: Optional[str] = None


class MealItemCreate(BaseModel):
    """Позиция для создания приёма пищи (КБЖУ на порцию)."""
    name: str
    amount_g: float = 100
    calories: float = 0
    protein_g: float = 0
    fat_g: float = 0
    carbs_g: float = 0


class MealCreateDto(BaseModel):
    """DTO создания приёма пищи."""
    meal_type: str = Field(..., pattern="^(breakfast|lunch|dinner|snack)$")
    eaten_at: Optional[datetime] = None  # если не указано — текущее время
    items: List[MealItemCreate]
    notes: str = ""


class MealUpdateDto(BaseModel):
    """DTO обновления позиций приёма пищи."""
    items: List[MealItemCreate]


class WaterLogDto(BaseModel):
    """DTO логирования воды."""
    amount_ml: int = Field(..., gt=0, le=5000)


class WaterOut(BaseModel):
    """Ответ при логировании воды."""
    id: int
    amount_ml: int
    logged_at: Optional[str] = None


class NutritionGoalOut(BaseModel):
    """Текущие цели по питанию."""
    calories: Optional[int] = None
    protein_g: Optional[int] = None
    fat_g: Optional[int] = None
    carbs_g: Optional[int] = None
    water_ml: int = 2000
    goal_type: Optional[str] = None         # lose / maintain / gain
    activity_level: Optional[str] = None    # sedentary / light / moderate / active / very_active


class NutritionGoalUpdate(BaseModel):
    """DTO обновления целей (передаются только изменяемые поля)."""
    calories: Optional[int] = None
    protein_g: Optional[int] = None
    fat_g: Optional[int] = None
    carbs_g: Optional[int] = None
    water_ml: Optional[int] = None
    goal_type: Optional[str] = None         # lose / maintain / gain
    activity_level: Optional[str] = None    # sedentary / light / moderate / active / very_active


class FoodItemWithFav(BaseModel):
    """Продукт с флагом избранного."""
    id: int
    name: str
    calories: Optional[float] = None
    protein_g: Optional[float] = None
    fat_g: Optional[float] = None
    carbs_g: Optional[float] = None
    serving_size_g: Optional[float] = None
    is_favorite: bool = False


class ProfileOut(BaseModel):
    """Параметры тела пользователя."""
    weight_kg: Optional[float] = None
    height_cm: Optional[float] = None
    age: Optional[int] = None
    gender: Optional[str] = None    # male / female


class ProfileUpdate(BaseModel):
    """DTO обновления параметров тела."""
    weight_kg: Optional[float] = None
    height_cm: Optional[float] = None
    age: Optional[int] = None
    gender: Optional[str] = None


class GoalsCalculateRequest(BaseModel):
    """Запрос на авто-расчёт целей по профилю."""
    goal_type: str = Field(..., description="lose / maintain / gain")
    activity_level: str = Field(..., description="sedentary / light / moderate / active / very_active")
    weight_kg: float = Field(..., gt=0)
    height_cm: float = Field(..., gt=0)
    age: int = Field(..., gt=0, le=120)
    gender: str = Field(..., description="male / female")
    water_ml: int = Field(default=2000, ge=0)


class TemplateItemDto(BaseModel):
    """Позиция шаблона для создания."""
    food_item_id: int
    amount_g: float


class TemplateCreateDto(BaseModel):
    """DTO создания шаблона."""
    name: str = Field(..., min_length=1, max_length=200)
    meal_type: str = "snack"
    items: list[TemplateItemDto]


class TemplateFromMealDto(BaseModel):
    """DTO создания шаблона из существующего приёма пищи."""
    meal_id: int
    name: str = Field(..., min_length=1, max_length=200)


class GoalsCalculateResponse(BaseModel):
    """Рассчитанные целевые КБЖУ."""
    calories: int
    protein_g: int
    fat_g: int
    carbs_g: int
    water_ml: int


class TotalsOut(BaseModel):
    """Суммарные КБЖУ за день."""
    calories: float = 0
    protein_g: float = 0
    fat_g: float = 0
    carbs_g: float = 0


class DaySummaryOut(BaseModel):
    """Полная сводка за день."""
    date: str
    goals: Optional[NutritionGoalOut] = None
    totals: TotalsOut
    water_ml: int = 0
    meals: List[MealOut] = []


class StatsOut(BaseModel):
    """Агрегированная статистика за период."""
    period: str
    days: int
    avg_calories: float
    avg_protein: float
    avg_fat: float
    avg_carbs: float
    avg_water: float
    daily: List[DaySummaryOut] = []


# ── Эндпоинты ─────────────────────────────────────────────────────────────────

@router.get("/today", response_model=DaySummaryOut)
async def get_today_summary(user: User = Depends(get_current_user)):
    """Сводка по питанию за сегодня."""
    today = date.today()
    # Делегируем в storage — возвращает готовый dict
    return await ns.get_nutrition_summary(user.telegram_id, today)


@router.get("/meals", response_model=List[MealOut])
async def get_meals(
    date_from: date = Query(..., description="Начало периода (YYYY-MM-DD)"),
    date_to: date = Query(..., description="Конец периода (YYYY-MM-DD)"),
    user: User = Depends(get_current_user),
):
    """Список приёмов пищи за период."""
    # Конвертируем date → datetime (начало/конец дня) с UTC
    dt_from = datetime.combine(date_from, datetime.min.time(), tzinfo=timezone.utc)
    dt_to = datetime.combine(date_to, datetime.max.time(), tzinfo=timezone.utc)
    return await ns.get_meals_by_date(user.telegram_id, dt_from, dt_to)


@router.get("/meal/{meal_id}", response_model=MealOut)
async def get_meal(meal_id: int, user: User = Depends(get_current_user)):
    """Приём пищи по ID."""
    result = await ns.get_meal(meal_id, user.telegram_id)
    if not result:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Приём пищи не найден")
    return result


@router.post("/meal", response_model=MealOut, status_code=status.HTTP_201_CREATED)
async def create_meal(dto: MealCreateDto, user: User = Depends(get_current_user)):
    """Создание приёма пищи (ручное из MiniApp)."""
    # Используем текущее время если не указано
    eaten_at = dto.eaten_at or datetime.now(timezone.utc)
    # Конвертируем Pydantic-модели в dict для storage
    items = [item.model_dump() for item in dto.items]
    return await ns.add_meal(
        user_id=user.telegram_id,
        meal_type=dto.meal_type,
        eaten_at=eaten_at,
        items=items,
        notes=dto.notes,
    )


@router.put("/meal/{meal_id}", response_model=MealOut)
async def update_meal(
    meal_id: int,
    dto: MealUpdateDto,
    user: User = Depends(get_current_user),
):
    """Обновление позиций приёма пищи."""
    items = [item.model_dump() for item in dto.items]
    result = await ns.update_meal_items(meal_id, user.telegram_id, items)
    if not result:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Приём пищи не найден")
    return result


@router.delete("/meal/{meal_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_meal(meal_id: int, user: User = Depends(get_current_user)):
    """Удаление приёма пищи."""
    deleted = await ns.delete_meal(meal_id, user.telegram_id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Приём пищи не найден")


@router.post("/water", response_model=WaterOut)
async def log_water(dto: WaterLogDto, user: User = Depends(get_current_user)):
    """Логирование потребления воды."""
    return await ns.add_water(user.telegram_id, dto.amount_ml)


@router.get("/water")
async def get_water(
    date: date = Query(default=None, description="Дата (YYYY-MM-DD), по умолчанию сегодня"),
    user: User = Depends(get_current_user),
):
    """Потребление воды за день."""
    target = date or __import__("datetime").date.today()
    total_ml = await ns.get_water_for_date(user.telegram_id, target)
    return {"date": target.isoformat(), "total_ml": total_ml}


@router.get("/stats", response_model=StatsOut)
async def get_stats(
    period: str = Query("week", pattern="^(week|month)$"),
    user: User = Depends(get_current_user),
):
    """Агрегированная статистика за неделю или месяц."""
    from datetime import timedelta
    today = date.today()
    # Определяем количество дней для агрегации
    days_count = 7 if period == "week" else 30

    # Собираем сводку за каждый день
    daily = []
    total_cal, total_p, total_f, total_c, total_w = 0, 0, 0, 0, 0
    for i in range(days_count):
        d = today - timedelta(days=i)
        summary = await ns.get_nutrition_summary(user.telegram_id, d)
        daily.append(summary)
        # Суммируем для средних
        total_cal += summary["totals"]["calories"]
        total_p += summary["totals"]["protein_g"]
        total_f += summary["totals"]["fat_g"]
        total_c += summary["totals"]["carbs_g"]
        total_w += summary["water_ml"]

    # Считаем средние
    return StatsOut(
        period=period,
        days=days_count,
        avg_calories=round(total_cal / days_count, 1),
        avg_protein=round(total_p / days_count, 1),
        avg_fat=round(total_f / days_count, 1),
        avg_carbs=round(total_c / days_count, 1),
        avg_water=round(total_w / days_count, 1),
        daily=daily,
    )


@router.get("/goals", response_model=NutritionGoalOut)
async def get_goals(user: User = Depends(get_current_user)):
    """Текущие цели по питанию."""
    goals = await ns.get_goals(user.telegram_id)
    if not goals:
        # Возвращаем дефолтные цели
        return NutritionGoalOut(calories=2000, protein_g=120, fat_g=65, carbs_g=250, water_ml=2000)
    return goals


@router.put("/goals", response_model=NutritionGoalOut)
async def update_goals(dto: NutritionGoalUpdate, user: User = Depends(get_current_user)):
    """Обновление целей по питанию."""
    return await ns.set_goals(
        user_id=user.telegram_id,
        calories=dto.calories,
        protein_g=dto.protein_g,
        fat_g=dto.fat_g,
        carbs_g=dto.carbs_g,
        water_ml=dto.water_ml,
        goal_type=dto.goal_type,
        activity_level=dto.activity_level,
    )


# ── Шаблоны приёмов пищи ───────────────────────────────────────────────────


@router.get("/templates")
async def get_templates(user: User = Depends(get_current_user)):
    """Список шаблонов пользователя."""
    return await ns.list_templates(user.telegram_id)


@router.post("/templates", status_code=status.HTTP_201_CREATED)
async def create_template(dto: TemplateCreateDto, user: User = Depends(get_current_user)):
    """Создать шаблон из списка продуктов."""
    items = [{"food_item_id": i.food_item_id, "amount_g": i.amount_g} for i in dto.items]
    return await ns.create_template(user.telegram_id, dto.name, dto.meal_type, items)


@router.post("/templates/from-meal", status_code=status.HTTP_201_CREATED)
async def create_template_from_meal(dto: TemplateFromMealDto, user: User = Depends(get_current_user)):
    """Создать шаблон из существующего приёма пищи."""
    result = await ns.create_template_from_meal(user.telegram_id, dto.meal_id, dto.name)
    if not result:
        raise HTTPException(status_code=404, detail="Приём пищи не найден")
    return result


@router.delete("/templates/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_template(template_id: int, user: User = Depends(get_current_user)):
    """Удалить шаблон."""
    deleted = await ns.delete_template(template_id, user.telegram_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Шаблон не найден")


@router.post("/templates/{template_id}/apply")
async def apply_template(template_id: int, user: User = Depends(get_current_user)):
    """Применить шаблон — создать приём пищи из шаблона."""
    result = await ns.apply_template(template_id, user.telegram_id)
    if not result:
        raise HTTPException(status_code=404, detail="Шаблон не найден")
    return result


# ── Смарт-повтор (предложения) ─────────────────────────────────────────────


@router.get("/suggestions")
async def get_suggestions(
    meal_type: Optional[str] = Query(None, description="breakfast/lunch/dinner/snack"),
    user: User = Depends(get_current_user),
):
    """Частые продукты и последние приёмы пищи для предложений.
    Если для конкретного meal_type данных нет — возвращаем общие (без фильтра)."""
    # Сначала ищем по конкретному типу приёма
    frequent = await ns.get_frequent_foods(user.telegram_id, meal_type=meal_type, limit=5)
    recent = await ns.get_recent_meals(user.telegram_id, meal_type=meal_type, limit=3)
    # Если по типу пусто — fallback на общие данные без фильтра
    if not frequent and meal_type:
        frequent = await ns.get_frequent_foods(user.telegram_id, meal_type=None, limit=5)
    if not recent and meal_type:
        recent = await ns.get_recent_meals(user.telegram_id, meal_type=None, limit=3)
    return {"frequent_foods": frequent, "recent_meals": recent}


# ── Избранные продукты ─────────────────────────────────────────────────────


@router.get("/favorites", response_model=List[FoodItemWithFav])
async def get_favorites(user: User = Depends(get_current_user)):
    """Список избранных продуктов пользователя."""
    return await ns.list_favorites(user.telegram_id)


@router.post("/favorites/{food_item_id}")
async def toggle_favorite(food_item_id: int, user: User = Depends(get_current_user)):
    """Добавить/убрать продукт из избранного (toggle)."""
    added = await ns.toggle_favorite(user.telegram_id, food_item_id)
    return {"food_item_id": food_item_id, "is_favorite": added}


# ── Профиль (параметры тела) ───────────────────────────────────────────────


@router.get("/profile", response_model=ProfileOut)
async def get_profile(user: User = Depends(get_current_user)):
    """Получить параметры тела пользователя."""
    profile = await ns.get_profile(user.telegram_id)
    if not profile:
        return ProfileOut()  # пустой профиль
    return profile


@router.put("/profile", response_model=ProfileOut)
async def update_profile(dto: ProfileUpdate, user: User = Depends(get_current_user)):
    """Обновить параметры тела пользователя."""
    return await ns.update_profile(
        user_id=user.telegram_id,
        weight_kg=dto.weight_kg,
        height_cm=dto.height_cm,
        age=dto.age,
        gender=dto.gender,
    )


@router.post("/goals/calculate", response_model=GoalsCalculateResponse)
async def calculate_goals(dto: GoalsCalculateRequest, user: User = Depends(get_current_user)):
    """
    Авто-расчёт целевых КБЖУ по параметрам тела и цели.
    НЕ сохраняет — возвращает рассчитанные значения для предпросмотра.
    """
    from services.nutrition_calc import calculate_full
    targets = calculate_full(
        weight_kg=dto.weight_kg,
        height_cm=dto.height_cm,
        age=dto.age,
        gender=dto.gender,
        activity_level=dto.activity_level,
        goal_type=dto.goal_type,
        water_ml=dto.water_ml,
    )
    return GoalsCalculateResponse(
        calories=targets.calories,
        protein_g=targets.protein_g,
        fat_g=targets.fat_g,
        carbs_g=targets.carbs_g,
        water_ml=targets.water_ml,
    )


@router.get("/food/search", response_model=List[FoodItemWithFav])
async def search_food(
    q: str = Query(..., min_length=1, max_length=100, description="Поисковый запрос"),
    user: User = Depends(get_current_user),
):
    """Поиск продуктов по названию."""
    results = await ns.search_food(q, user_id=user.telegram_id)
    # Добавляем флаг is_favorite к результатам
    fav_ids = await ns.get_favorite_ids(user.telegram_id)
    return [
        {**r, "is_favorite": r["id"] in fav_ids}
        for r in results
    ]


# ── Score / Remaining / Weekly Summary ────────────────────────────────────────

@router.get("/score")
async def get_daily_score(
    target_date: Optional[date] = Query(None, description="Дата в формате YYYY-MM-DD (по умолчанию — сегодня)"),
    user: User = Depends(get_current_user),
):
    """Оценка дня по питанию (score 0-100) с разбивкой по компонентам."""
    from services.nutrition_score import calculate_daily_score
    d = target_date or date.today()
    return await calculate_daily_score(user.telegram_id, d)


@router.get("/remaining")
async def get_remaining_today(
    user: User = Depends(get_current_user),
):
    """Остаток КБЖУ на сегодня (разница между целями и съеденным)."""
    today = date.today()
    summary = await ns.get_nutrition_summary(user.telegram_id, today)
    goals = summary.get("goals") or {}
    totals = summary.get("totals", {})
    water_ml = summary.get("water_ml", 0)

    if not goals or not goals.get("calories"):
        raise HTTPException(status_code=400, detail="Цели по КБЖУ не установлены.")

    return {
        "date": str(today),
        "remaining": {
            "calories": max(0, goals["calories"] - totals.get("calories", 0)),
            "protein_g": max(0, (goals.get("protein_g") or 0) - totals.get("protein_g", 0)),
            "fat_g": max(0, (goals.get("fat_g") or 0) - totals.get("fat_g", 0)),
            "carbs_g": max(0, (goals.get("carbs_g") or 0) - totals.get("carbs_g", 0)),
            "water_ml": max(0, (goals.get("water_ml") or 0) - water_ml),
        },
        "eaten": totals,
        "goals": goals,
        "progress_pct": round(totals.get("calories", 0) / goals["calories"] * 100) if goals["calories"] else 0,
    }


@router.get("/weekly-summary")
async def get_weekly_summary(
    user: User = Depends(get_current_user),
):
    """Недельный AI-обзор питания с рекомендациями."""
    from services.nutrition_weekly_summary import generate_weekly_summary
    return await generate_weekly_summary(user.telegram_id)
